tensorboard/data/server/logdir.rs

```diff
@@ -246,22 +263,27 @@ impl<'a> LogdirLoader<'a> {
             .runs
             .read()
             .expect("could not acquire runs.data");
+
+        let mut work_items = Vec::new();
         for (run, run_state) in self.runs.iter_mut() {
             let event_files = discoveries
                 .remove(run)
                 .unwrap_or_else(|| panic!("run in self.runs but not discovered: {:?}", run));
             let filenames: Vec<PathBuf> = event_files.into_iter().map(|d| d.event_file).collect();
-            run_state.loader.reload(
-                filenames,
-                commit_runs.get(run).unwrap_or_else(|| {
-                    panic!(
-                        "run in self.runs but not in commit.runs \
+            let run_data = commit_runs.get(run).unwrap_or_else(|| {
+                panic!(
+                    "run in self.runs but not in commit.runs \
                         (is another client mutating this commit?): {:?}",
-                        run
-                    )
-                }),
-            );
+                    run
+                )
+            });
+            work_items.push((&mut run_state.loader, filenames, run_data));
```

Comment:
I might just be tying myself in knots here, but do you have any thoughts for how I can better understand the logic the borrow checker uses to handle this kind of case where we mutably borrow each `run_state.loader` and push them all into another vector for later extraction?  I guess the original `self.runs` lifetime is the outer bound for the individual loaders, and then the `Vec` inherits their lifetime, but then we take the loaders back out of `Vec` - and I guess the lifetime follows them through to wherever they go?  But presumably it's also somehow attached to the `Vec` itself now, right, so even if we were to empty the `Vec` entirely and then try to go refill it without dropping it in between, that would still prevent mutation of `self.runs` even though the actual loader references are all gone?

Reply:
Yep. Let’s spelunk through the logic:

  - This method takes `&mut self`; call that elided lifetime `'1`, so
    it’s `fn load_runs<'1>(&'1 mut self, ...)`.
  - We borrow `self.runs.iter_mut()` for lifetime `'2`, which must not
    outlive `'1` so that it doesn’t outlive borrowed content.
  - Then we borrow `&mut run_state.loader`, for lifetime `'3`, which
    must not outlive `'3` so that it doesn’t outlive *that* loan.
  - These are the lifetime upper bounds: `'1: '2` and `'2: '3`, where
    “`:`” is a subtyping relation that you can pronounce “outlives (or
    is equal to)”. You’ll often see `T: 'static` to indicate that `T`
    has only static references (“outlives everything”).
  - We have a `Vec::<(&'v mut RunLoader, ...)>` for some lifetime `'v`.
    Since we `push` items with lifetime `'3`, we must have `'3: 'v`: the
    referenced content must outlive the vector.
  - Now, we move `work_items` into the closure arg to `install`. From
    here on out, there aren’t really any more lifetimes: we call
    `into_par_iter` and `for_each`, both of which are simply typed
    methods with by-value receivers. In particular, neither of them has
    a lifetime bound on its closure type, unlike `std::thread::spawn`.

You can similarly analyze the lifetime flow for the `&RwLock<...>`
lifetime. You’ll get something like this:

![Lifetime graph; for details, see https://gist.github.com/wchargin/e18bca00fbf7f2eb6ae77a193fc2048e][graph]

(Here, an arrow from `a` to `b` means `a: b`.)

[graph]: https://user-images.githubusercontent.com/4317806/106556041-9411be80-64d3-11eb-9825-fc7eabb9d07b.png

Writing it out, the inference looks pretty unidirectional here: we
require `'1: '2: '3: 'v`, but there’s nothing concrete at the end of
that chain, so this is satisfiable by (e.g.) setting `'1 = '2 = '3 = 'v`
and the borrow checker is happy.

If we had called [`std::thread::spawn`] instead of `install`, we would
then require `'v: 'static`, since `spawn` requires that `F: 'static`
(where `F` is the type of the closure), and we have `'v: F` since the
vector must live as long as the closure itself. This transitively means
that the whole diagram must outlive `'static`. But `'1` was arbitrary:
it’s the lifetime to the receiver parameter. Thus, this isn’t
satisfiable in general and should be a borrowck error.

Let’s see what Rustc has to say. In this patch, I’ve dropped the
`run_data` arg so that the compiler is forced to complain about the left
branch of the graph above rather than the right branch (it’s the same
logic, just not the one that you asked about):

```diff
diff --git a/tensorboard/data/server/logdir.rs b/tensorboard/data/server/logdir.rs
--- a/tensorboard/data/server/logdir.rs
+++ b/tensorboard/data/server/logdir.rs
@@ -277,12 +277,13 @@ impl<'a> LogdirLoader<'a> {
                     run
                 )
             });
-            work_items.push((&mut run_state.loader, filenames, run_data));
+            work_items.push((&mut run_state.loader, filenames));
         }
-        self.thread_pool.install(|| {
-            work_items
-                .into_par_iter()
-                .for_each(|(loader, filenames, run_data)| loader.reload(filenames, run_data));
+        drop(commit_runs);
+        std::thread::spawn(move || {
+            for (loader, filenames) in work_items.into_iter() {
+                loader.reload(filenames, todo!("omitted for example"));
+            }
         });
     }
 }
```

Rustc says:

```
error[E0759]: `self` has an anonymous lifetime `'_` but it needs to satisfy a `'static` lifetime requirement
   --> logdir.rs:268:43
    |
259 |     fn load_runs(&mut self, discoveries: Discoveries) {
    |                  --------- this data with an anonymous lifetime `'_`...
...
268 |         for (run, run_state) in self.runs.iter_mut() {
    |                                 --------- ^^^^^^^^
    |                                 |
    |                                 ...is captured here...
...
283 |         std::thread::spawn(move || {
    |         ------------------ ...and is required to live as long as `'static` here
```

This is pointing out the transitive `'1: '2: 'static` chain and
complaining about its unsatisfiability.

So how can Rayon’s `install` or Crossbeam’s `scope` get away with this
while `thread::spawn` needs the `'static` requirement? The key is that
the scoped thread implementations *ensure* that the callback finishes
before `install`/`scope` itself returns. Thus, the borrowed content only
needs to live as long as the call to `install`, which is on the caller’s
stack frame and is thus known to be outlived by `'1`. Under the hood,
both of these will call `thread::spawn` after [using `unsafe` to
transmute your closure to `'static`][tm]. (If “unsafe” is surprising
here, remember that *soundness* is what we care about. “Sound” means “in
no real execution will it trigger undefined behavior”; “safe” means “the
compiler has statically proven that it is sound”; “unsafe” means “*the
code author* has statically proven that it is sound based on properties
too advanced to be represented in the base type system. We use sound
`unsafe` code to build safe abstractions.)

[tm]: https://docs.rs/crossbeam-utils/0.8.0/src/crossbeam_utils/thread.rs.html#445-451

> But presumably it's also somehow attached to the `Vec` itself now, right, so even if we were to empty the `Vec` entirely and then try to go refill it without dropping it in between, that would still prevent mutation of `self.runs` even though the actual loader references are all gone?

Yeah, definitely. You can get that kind of flexibility with [`RefCell`],
which basically pushes the borrow checking to runtime. (Follows a Rust
pattern: you can have dynamic borrow checking, dynamic dispatch, dynamic
reference counting, dynamic typing (`std::any`), but you have to opt in,
and sometimes accept consequences like “your type is no longer `Sync`”.)
But a humble `Vec<_>` isn’t that smart by default.

[`RefCell`]: https://doc.rust-lang.org/std/cell/struct.RefCell.html
[`std::any`]: https://doc.rust-lang.org/std/any/index.html
[`std::thread::spawn`]: https://doc.rust-lang.org/std/thread/fn.spawn.html

Does this help?

Reply:
Got it, that helps!  Thanks very much for the thorough explanation.